hw/xbox/nvnet.c

```diff
@@ -19,78 +20,983 @@
+            break;
+        }
+        if (val & NVREG_TXRXCTL_BIT1)
+        {
+            nvnet_set_reg(s, NvRegIrqStatus, 0, 4);
+            break;
+        }
+        else if (val == 0)
+        {
+            temp = nvnet_get_reg(s, NvRegUnknownSetupReg3, 4);
+            if (temp == NVREG_UNKSETUP3_VAL1) {
+                /* forcedeth waits for this bit to be set... */
+                nvnet_set_reg(s, NvRegUnknownSetupReg5,
+                                 NVREG_UNKSETUP5_BIT31, 4);
+                break;
+            }
+        }
+
+        nvnet_set_reg(s, NvRegTxRxControl, val, size);
+
+        break;
+
+    case NvRegIrqMask:
+        nvnet_set_reg(s, addr, val, size);
+        nvnet_update_irq(s);
+        break;
+
+    case NvRegIrqStatus:
+        nvnet_set_reg(s, addr, nvnet_get_reg(s, addr, size) & ~val, size);
+        nvnet_update_irq(s);
+        break;
+
+    default:
+        nvnet_set_reg(s, addr, val, size);
+        break;
+    }
 }
+
 static const MemoryRegionOps nvnet_mmio_ops = {
     .read = nvnet_mmio_read,
     .write = nvnet_mmio_write,
 };
 
+/*******************************************************************************
+ * Packet TX/RX
+ ******************************************************************************/
 
-static uint64_t nvnet_io_read(void *opaque,
-                              hwaddr addr, unsigned int size)
+static void nvnet_send_packet(NvNetState *s, const uint8_t *buf, int size)
+{
+    NetClientState *nc = qemu_get_queue(s->nic);
+
+    NVNET_DPRINTF("nvnet: Sending packet!\n");
+    nvnet_hex_dump(s, buf, size);
+    qemu_send_packet(nc, buf, size);
+}
+
+static int nvnet_can_receive(NetClientState *nc)
+{
+    //NvNetState *s = qemu_get_nic_opaque(nc);
+    NVNET_DPRINTF("nvnet_can_receive called\n");
+    return 1;
+}
+
+static ssize_t nvnet_receive(NetClientState *nc,
+                             const uint8_t *buf, size_t size)
 {
-    NVNET_DPRINTF("nvnet IO: read [0x%llx]\n", addr);
+    const struct iovec iov = {
+        .iov_base = (uint8_t *)buf,
+        .iov_len = size
+    };
+
+    NVNET_DPRINTF("nvnet_receive called\n");
+    return nvnet_receive_iov(nc, &iov, 1);
+}
+
+static ssize_t nvnet_receive_iov(NetClientState *nc,
+                                 const struct iovec *iov, int iovcnt)
+{
+    NvNetState *s = qemu_get_nic_opaque(nc);
+    size_t size = iov_size(iov, iovcnt);
+
+    NVNET_DPRINTF("nvnet: Packet received!\n");
+
+    if (size > sizeof(s->txrx_dma_buf))
+    {
+        NVNET_DPRINTF("nvnet_receive_iov packet too large!\n");
+        assert(0);
+        return -1;
+    }
+
+    iov_to_buf(iov, iovcnt, 0, s->txrx_dma_buf, size);
+    nvnet_hex_dump(s, s->txrx_dma_buf, size);
+    return nvnet_dma_packet_to_guest(s, s->txrx_dma_buf, size);
+}
+
+static ssize_t nvnet_dma_packet_to_guest(NvNetState *s,
+                                         const uint8_t *buf, size_t size)
+{
+    struct RingDesc desc;
+
+    for (int i = 0; i < s->rx_ring_size; i++)
```

Comment:
nit: not picked up by checkpatch, but no qemu code seems to use inline for initialisers for some reason